### ISSUE TITLE
Feature module - add translations instead of overwriting the existing ones

### ISSUE THREAD

**Issue by Etchelon** [robisim74/angular-l10n#22]
Ciao di nuovo,
I'm playing with your library and I've noticed that addTranslation/translationProvider will overwrite the translationData object in LocalizationService instead of adding the new translations.
More details:
I have a SharedModule that exports a singleton LocalizationService in the forRoot method (as explained in the angular documentation) because this SharedModule is also imported by my lazy loaded module, which I'll call BModule.
In the constructor of BModule (and not in the constructor of the main component in BModule, since there's no single entry point but depends on the route!) I wrote this:
`export default class BModule {
    constructor(_localizationService: LocalizationService) {
        //_localizationService.translationProvider("/app/ehr/resources/");
        _localizationService.addTranslation("it", {
            "SET_STATUS": "Imposta stato a:"
        });
        _localizationService.updateTranslation();
    }
}`
and after the addTranslation/translationProvider calls the translationData object is overwritten. I don't want this because I want to keep global translations (generic things like save, cancel, ecc...) in the "core module" of my webapp, and load the translations specific for BModule and its components afterwards.
Therefore, coupling this behavior with namespaced keys (see my other Issue) I can write, in the template of a component in the BModule, something like:
`<div>
  <h1>{{ "BCOMPONENT1.TITLE" | translate:lang }}</h1>
  <button type="button" class="btn btn-success">{{ CORE.OK | translate:lang }}</button>
  <button type="button" class="btn btn-danger">{{ CORE.CANCEL | translate:lang }}</button>
</div>`

Memory usage will still be optimized: BModule resources will be loaded ONLY if the user navigates to a proper route, otherwise not!

**Comment by robisim74** [comment]
Ciao di nuovo Andrea,

if you export `LocalizationModule.forRoot()` in a shared module, you create an instance of  `LocalizationService` for each loaded lazy module.
This does not mean that the translation data is overwritten, but you must provide new translation data to the new instance, as explained [here](https://github.com/robisim74/angular2localization/blob/master/doc/spec.md#5) and as is shown in the [sample app](https://github.com/robisim74/angular2localization/blob/gh-pages/app/list.module.ts). 
This new instance replaces the previous one and is inherited by all its children components.
All this is correct: it is the mechanism of _Dependency Injection_.

You say: _I don't want this because I want to keep global translations (generic things like save, cancel, ecc...) in the "core module" of my webapp_. Ok, you have two ways to do it:
1. Creating a json file for each language, shared by the whole application: in this case you do not export the `LocalizationModule` into the shared module, but in the core module (it will be singleton). See [here](https://angular.io/docs/ts/latest/guide/ngmodule.html#!#core-module).
2. Creating your own `GlobalLocalizationService` that _extends_ `LocalizationService`, and initializing it with the global translations like save, cancel etc. At this point, you can also create a new `GlobalTranslatePipe`, which _extends_ `TranslatePipe`, and pass it `GlobalLocalizationService` rather than `LocalizationService`: the new pipe will be used in the global translations: `{{'SAVE' | globalTranslate: lang}}`. Of course, then you have to export the new service and the new pipe in the core module.

Basically, the library provides classes: find you the way to use them according to the need. I hope this is helpful.

Greetings,

Roberto

**Comment by Etchelon** [comment]
Ciao,

I explained myself wrong. I already export the LocalizationService as a singleton:

``` typescript
    @NgModule({
    imports: [
        CommonModule,
        FormsModule
    ],
    exports: [
        CommonModule,
        FormsModule,
        LocaleModule,
        LocalizationModule,
        SelectMultipleComponent
    ],
    declarations: [SelectMultipleComponent],
    providers: [
        AuthGuard,
        SelectedOuGuard
    ]
    })
    export class SharedModule {
        static forRoot(): ModuleWithProviders {
            return {
                ngModule: SharedModule,
                providers: [
                    AuthService,
                    LocalizationService
                ]
            };
        }
    }
```

With the static forRoot, I provide AuthService (as explained in angular's docs) and LocalizationService as singletons, while LocalizationModule.forRoot() is in the imports of the core module of the app.
In fact, if I didn't, I'd have a new instance of LocalizationService for BModule when I load it.

But it's a singleton and in fact, I start my app, load the core resources, and see my main page localized. Then load BModule, and in its constructor I see the singleton instance get injected, with all the core translations in the translationData property. This is what I want! What I'd also like is to add new resources to the singleton:

``` typescript
    @NgModule({
        imports: [SharedModule, routing],
        declarations: [*My stuff*]
    })
    export default class BModule {
        constructor(_localizationService: LocalizationService) {
            // HERE, _localizationService.translationData contains core resources
            _localizationService.addTranslation("it", {
                "SET_STATUS": "Imposta stato a:"
            });
            // HERE, _localizationService.translationData only contains "SET_STATUS"!! I wanted to add to the other resources :/
            _localizationService.updateTranslation();
        }
    }
```

I think it's more neat than creating another service for just the global resources.

I see an analogy with the new AddApplicationPart api in AspNet Core: the main WebApp defines a set of controllers, then I develop plugins to the app separately. These plugins add more controllers, but when I load them at startup, the additional controllers in the plugin's dll are merged into the main app. I hope I made myself clear with this analogy.

Greetings,

Andrea

**Comment by robisim74** [comment]
Ok , it's clear. You can not currently do this. I could do that translations are added rather than overwritten, but we should keep track of all loaded translations, we should check if a translation has already been loaded and we should do an http get for all loaded translations when the user changes language. I fear that this can slow down the application. I have to do some testing before.

**Comment by robisim74** [comment]
Ciao Andrea,

there is another problem we can not solve: if we add the data, the pure pipes can not see them, because they are updated only when there is a change in the input parameters (in our case the language).

So I had another idea: I added the possibility to supply more than one provider for each translation.

Essentially:
- Use the Lazy loading as shown [here](https://github.com/robisim74/angular2localization/blob/master/doc/spec.md#5);
- Add to initialization of each new instance of `LocalizationService` your provider of shared keys, for example:

```
// Initializes LocalizationService: asynchronous loading.
this.localization.addProvider('./resources/locale-list-');
this.localization.addProvider('./resources/global-'); // Contains the shared keys.
this.localization.updateTranslation(); // Need to update the translation.
```

I added the commit that allows you to use this feature.
If you want to test it, you can do so:
- Uninstall this library from your application
- Download this repository
- Follow the instructions of the README file to build the library and install it in your application

Please let me know if it works as you expect.

Saluti,

Roberto

**Comment by Etchelon** [comment]
I'll let you know tomorrow! Thanks for now :)

**Comment by robisim74** [comment]
Don't worry, I have a job, a young son and I'm studying at university... ;)

**Comment by Etchelon** [comment]
Wow :)
Anyway I stayed a bit late at work and tried it. Seems good so far, but only in the next days I'll start using the library more extensively and be able to give a more thorough opinion!
Thanks again ;)

Andrea

**Comment by Etchelon** [comment]
Ciao Roberto,
the fix you provided seems good to me. It doesn't really make all resources available everywhere, across all modules, but it's ok, my use case was to reuse the common resources from the core module inside feature modules. Also I won't have more than 2 feature modules open simultaneously, so it doesn't really matter.

Can you push it to npm?

p.s.: it appears you're using tsc 2.0.2, which generated a locale.d.ts file which takes advantage of the "readonly" keyword. Not a problem in compilation, just a few red lines in the output.

Thanks,

Andrea

**Comment by robisim74** [comment]
Ok, today I can't. I will publish tomorrow night.

Yes, I had to switch to TypeScript 2 for AoT compilation (Angular 2 uses TS 2).

**Comment by robisim74** [comment]
I released v1.0.2.

Greetings
